### PR TITLE
Add notations `\top` and `\bot` and deprecate `0` and `1`

### PR DESCRIPTION
**Pull request by proux01**
##### Motivation for this change

When using order in combination with ssralg, it is convenient to
```Coq
Local Open Scope ring_scope.
Local Open Scope order_scope.
```
except that now `0` and `1` mean top and bottom of lattices instead of neutral element of ring.

##### Things done/to do

- [x] added corresponding entries in `CHANGELOG_UNRELEASED.md` (do not edit former entries)
- [x] added corresponding documentation in the headers

##### Automatic note to reviewers

Read [this Checklist](https://github.com/math-comp/math-comp/wiki/Checklist-for-following,-reviewing-and-playing-with-a-PR#checklist-for-reviewing-a-pr) and make sure there is a milestone.

### CODE REVIEW

**Comment by proux01** [comment]
>     * I think that `order_scope` should not be opened unless you are reasoning about abstract order instances.
> 
>     * Opening `ring_scope` after `order_scope` seems better in general since it should give precedence to the notations in `ring_scope` in parsing:
>       > If a notation is defined in multiple scopes, Coq uses the interpretation from the most recently opened notation scope or declared lonely notation.

But then , how would you handle two concrete orders, one being on some numDomainType but not the other?

**Comment by pi8027** [comment]
> But then, how would you handle two concrete orders, one being on some numDomainType but not the other?

In that case, I would open only `ring_scope` and explicitly write `%O` (or another scope delimiter if applicable) for the other.

**Comment by proux01** [comment]
That sounds pretty inconvenient, why shouldn't we open `order_scope`? Is there something bad with having a generic `_ <= _` notation ofr all porder types? or some other reason?

**Comment by proux01** [comment]
@pi8027 I have given a bit more though to this and I realize maybe I'm just messing something up with my displays. Here is my use case:
* orders on `numDomainType` have a notation on `ring_display` that  parses and prints as `(_ <= _)%R`
* in Analysis (c.f. [constructive_ereal.v](https://github.com/math-comp/analysis/blob/master/theories/constructive_ereal.v)), the extended reals have a notation on `ereal_display` that also parses and prints as `(_ <= _)%E`

Maybe we should just reuse `ring_display` in `constructive_ereal.v` as we don't really want the display to differ from the one of `ring_display` after all? Doing this, I wouldn't have to open `order_scope` (or put `%E` scope annotations everywhere) but only `ring_scope` in my usecase.

**Comment by pi8027** [comment]
@proux01 From my superficial understanding, it is safer to have both `ring_display` (resp. `ring_scope`) and `ereal_display` (resp. `ereal_scope`). But `constructive_ereal.v` seems to lack specialized notations for `bottom` and `top` (and thus both `bottom` and `GRing.zero` are printed as `0`?). What about redefining `+oo` and `-oo` as such?

**Comment by proux01** [comment]
> `constructive_ereal.v` seems to lack specialized notations for `bottom` and `top` (and thus both `bottom` and `GRing.zero` are printed as `0`?). What about redefining `+oo` and `-oo` as such?

I agree, thanks for pointing this out.

> @proux01 From my superficial understanding, it is safer to have both ring_display (resp. ring_scope) and ereal_display (resp. ereal_scope).

Well, what's the point of having two different displays with the same... display? When writing pen and paper proofs I write `_ <= _` both for reals and extended reals, I would find it nice if I didn't need to change one to `(_ <= _)%R` or `(_ <= _)%E` when formalizing.

**Comment by proux01** [comment]
Could you elaborate on your idea? I just don't grasp it unfortunately.

**Comment by pi8027** [comment]
Sure. `(x%:E <= y%:E)%E` should be convertible with `(x <= y)%R`, but we have to unfold `Order.le` in the first term to see this fact. IIUC, keyed matching does not unfold the head symbol to see this kind of fact (it checks that the arguments are respectively convertible). So it seems that ssr tactics may not find the right subterm.

**Comment by pi8027** [comment]
To be clear, I'm not against the changes introduced by this PR, but saying that there is a discipline in the use of displays, notation scopes, etc. We can open an issue on the analysis repo and continue the discussion.

**Comment by CohenCyril** [comment]
@pi8027 I agree we should have a more general discussion, but I also think the notations `0` and `1` were particularly bad in the first place :sweat_smile: 

@proux01 wow you went a long way to preserve backward compatiblity, but I'm not sure it's desirable, and it's definitely error prone since, unless I'm mistaken, you missed a few rescoping in the notations you added (e.g. ``x `&` y`` in `orderBT_scope` does not rescope `x` and `y` properly and it's not the only the only place).
I would be more aggressive and add `\bot` and `\top` and remove `0` and  `1` from `order_scope`, and risk breaking compat and add declarations in the few (I expect) places where it's really necessary.
We could also add unscoped `0%O` and `1%O` parsing only declarations, maybe we can even hack the display to trigger a deprecation warning?

**Comment by proux01** [comment]
> @pi8027 I agree we should have a more general discussion

I added the point to the agenda of the next Analysis meeting (extended reals are still there currently)

> I would be more aggressive and add `\bot` and `\top` and remove `0` and `1` from `order_scope`, and risk breaking compat and add declarations in the few (I expect) places where it's really necessary.

Indeed, done that

> We could also add unscoped `0%O` and `1%O` parsing only notations,

Done, this doesn't work with Coq 8.13 and 8.14 (apparently something got fixed/improved in 8.15), so I think this is a good time to drop support for 8.13 and 8.14, if we agree CI is now green.

> maybe we can even hack the display to trigger a deprecation warning?

Didn't manage to get anything along this line but since the notation is parsing only, the printing should already be an indice that something changed (beside it doesn't seem to be much used (apparently only mczify in the CI)).

**Comment by pi8027** [comment]
Just in case, I opened a PR to fix mczify math-comp/mczify#46.

**Comment by proux01** [comment]
Thanks, no hurry, that'll only be actually required when we'll remove the new deprecated `0%O` and `1%O` notations.

**Review by CohenCyril — APPROVED** [review]
LGTM

File: CHANGELOG_UNRELEASED.md [hunk 1]
```diff
@@ -45,6 +45,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 - in `ssrfun.v`
   + lemmas `inj_omap`, `omap_id`, `eq_omap`, `omapK`
+    `addBnA`, `subBnAC`
```
Comment (pi8027) [inline CHANGELOG_UNRELEASED.md:48]: This line is from another PR.

File: CHANGELOG_UNRELEASED.md [hunk 2]
```diff
@@ -45,6 +45,13 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
...
+- in `order.v`
+  + notations `0%O`, `1%O`, `0^d%O` and `1^d%O` as backward compatible
+    replacements of removed  notation `0`, `1`, `0^d` and `1^d`
+    for bottom and top of lattices
+  + notations `\top` and `\bot` for `Order.top` and `Order.bottom`
+  + notations `\top^d` and `\bot\d` for `Order.dual_top` and `Order.dual_bottom`
```
Comment (pi8027) [inline CHANGELOG_UNRELEASED.md:54]:
```suggestion
  + notations `\top^d` and `\bot^d` for `Order.dual_top` and `Order.dual_bottom`
```

File: CHANGELOG_UNRELEASED.md (hunk 1 above)
Comment (proux01) [reply CHANGELOG_UNRELEASED.md]: Indeed, thanks!

**Pull request merged**